editor/src/core/workers/parser-printer/parser-printer-parsing.ts

```diff
@@ -1165,47 +1214,154 @@ function createExpressionOtherJavaScript(
+  return withParserMetadata(
+    expression,
+    buildHighlightBoundsForUids(sourceFile, node, uid),
+    propsUsed,
+    definedElsewhere,
+  )
 }
 
 function createExpressionNestedObject(
   sourceFile: TS.SourceFile,
+  node: TS.Node,
   objectContents: Array<JSXProperty>,
   comments: ParsedComments,
   alreadyExistingUIDs: Set<string>,
-): JSExpressionNestedObject {
+  propsUsed: Array<string>,
+  definedElsewhere: Array<string>,
+): WithParserMetadata<JSExpressionNestedObject> {
   const value = jsExpressionNestedObject(objectContents, comments, '')
   const uid = generateUIDAndAddToExistingUIDs(sourceFile, value, alreadyExistingUIDs)
-  return jsExpressionNestedObject(objectContents, comments, uid)
+  const expression = jsExpressionNestedObject(objectContents, comments, uid)
+  return withParserMetadata(
+    expression,
+    buildHighlightBoundsForUids(sourceFile, node, uid),
+    propsUsed,
+    definedElsewhere,
+  )
 }
 
 function createExpressionFunctionCall(
   sourceFile: TS.SourceFile,
+  node: TS.Node,
   functionName: string,
   parameters: Array<JSExpression>,
   alreadyExistingUIDs: Set<string>,
-): JSExpressionFunctionCall {
+  propsUsed: Array<string>,
+  definedElsewhere: Array<string>,
+): WithParserMetadata<JSExpressionFunctionCall> {
   const value = jsExpressionFunctionCall(functionName, parameters, '')
   const uid = generateUIDAndAddToExistingUIDs(sourceFile, value, alreadyExistingUIDs)
-  return jsExpressionFunctionCall(functionName, parameters, uid)
+  const expression = jsExpressionFunctionCall(functionName, parameters, uid)
+  return withParserMetadata(
+    expression,
+    buildHighlightBoundsForUids(sourceFile, node, uid),
+    propsUsed,
+    definedElsewhere,
+  )
+}
+
+function createArbitraryJSBlock(
+  sourceFile: TS.SourceFile,
+  javascript: string,
+  transpiledJavascript: string,
+  definedWithin: Array<string>,
+  definedElsewhere: Array<string>,
+  sourceMap: RawSourceMap | null,
+  elementsWithin: ElementsWithin,
+  alreadyExistingUIDs: Set<string>,
+): ArbitraryJSBlock {
+  const value = arbitraryJSBlock(
+    javascript,
+    transpiledJavascript,
+    definedWithin,
+    definedElsewhere,
+    sourceMap,
+    elementsWithin,
+    '',
+  )
+  const uid = generateUIDAndAddToExistingUIDs(sourceFile, value, alreadyExistingUIDs)
+  return arbitraryJSBlock(
+    javascript,
+    transpiledJavascript,
+    definedWithin,
+    definedElsewhere,
+    sourceMap,
+    elementsWithin,
+    uid,
+  )
+}
+
+function createJSXArbitraryBlock(
+  sourceFile: TS.SourceFile,
+  originalJavascript: string,
+  javascript: string,
+  transpiledJavascript: string,
+  definedElsewhere: Array<string>,
+  sourceMap: RawSourceMap | null,
+  elementsWithin: ElementsWithin,
+  alreadyExistingUIDs: Set<string>,
+): JSXArbitraryBlock {
+  // Ideally the value we hash is stable regardless of location, so exclude the SourceMap value from here and provide an empty UID.
```

Comment:
Should this also be true for `createArbitraryJSBlock` above?

Reply:
Yes, yes I think it should.